mindsdb/api/http/namespaces/predictor.py

```diff
@@ -290,6 +293,131 @@ def post(self, name):
+            'mariadb': MariaDS,
+            'mysql': MySqlDS,
+            'postgres': PostgresDS,
+            'mssql': MSSQLDS,
+            'mongodb': MongoDS,
+            'snowflake': SnowflakeDS
+        }
+
+        try:
+            dsClass = ds_class_map[integration['type']]
+        except KeyError:
+            raise KeyError(f"Unknown DS type: {integration_id}, type is {integration['type']}")
+
+        if integration['type'] in ['clickhouse']:
+            picklable = {
+                'class': dsClass.__name__,
+                'args': [],
+                'kwargs': {
+                    'query': data['query'],
+                    'user': integration['user'],
+                    'password': integration['password'],
+                    'host': integration['host'],
+                    'port': integration['port']
+                }
+            }
+            ds = dsClass(**picklable['kwargs'])
+
+        elif integration['type'] in ['mssql', 'postgres', 'mariadb', 'mysql']:
+            picklable = {
+                'class': dsClass.__name__,
+                'args': [],
+                'kwargs': {
+                    'query': data['query'],
+                    'user': integration['user'],
+                    'password': integration['password'],
+                    'host': integration['host'],
+                    'port': integration['port']
+                }
+            }
+
+            if 'database' in integration:
+                picklable['kwargs']['database'] = integration['database']
+
+            if 'database' in data:
+                picklable['kwargs']['database'] = data['database']
+
+            ds = dsClass(**picklable['kwargs'])
+
+        elif integration['type'] == 'snowflake':
+            picklable = {
+                'class': dsClass.__name__,
+                'args': [],
+                'kwargs': {
+                    'query': data['query'],
+                    'schema': data['schema'],
+                    'warehouse': data['warehouse'],
+                    'database': data['database'],
+                    'host': integration['host'],
+                    'password': integration['password'],
+                    'user': integration['user'],
+                    'account': integration['account']
+                }
+            }
+
+            ds = dsClass(**picklable['kwargs'])
+
+        elif integration['type'] == 'mongodb':
+            picklable = {
+                'class': dsClass.__name__,
+                'args': [],
+                'kwargs': {
+                    'database': data['database'],
+                    'collection': data['collection'],
+                    'query': data['find'],
+                    'user': integration['user'],
+                    'password': integration['password'],
+                    'host': integration['host'],
+                    'port': integration['port']
+                }
+            }
+
+            ds = dsClass(**picklable['kwargs'])
```

Comment:
A lot of this is repeat logic that is contained elsewhere (i.e. it's already done by the component used to create a datasource from a sql query).

It would be better to use the logic that's used to create a datasource from a sql query (maybe seprarated in it's own function if it currently isn't) and then separate the predict logic into another function, and then couple them into this endpoint.

Otherwise we end up with too much alsmot-copy-paste code that needs maintaining,

Reply:
Hi! We ended up separating the logic for creating a datasource from a sql query into a function in the DataStore class, as the code came from the save_datasource function. Let us know if there's a better place to put that function, as it doesn't really have much to do with the data storage system tbh.

Also, the prediction logic was separated into a new function in the same predictor.py file, and we've changed the three predict endpoints to use that function instead of repeating the same predict logic.